### PR TITLE
Kuznetsov Artyom lab2 (option 1)

### PR DESCRIPTION
**Pull request by Kuznetsov-Artyom**
Implemented a plugin that inserts the `instruction_start()`/`insrtuction_end()` functions at the beginning and end of the function, respectively.

Example:
```cpp
// test.cpp
int factorial(int value) {
  return value <= 1 ? 1 : value * factorial(value - 1);
}
```
```ll
; test.ll
define dso_local noundef i32 @_Z9factoriali(i32 noundef %value) {
entry:
  %value.addr = alloca i32, align 4
  store i32 %value, ptr %value.addr, align 4
  %0 = load i32, ptr %value.addr, align 4
  %cmp = icmp sle i32 %0, 1
  br i1 %cmp, label %cond.true, label %cond.false

cond.true:
  br label %cond.end

cond.false:
  %1 = load i32, ptr %value.addr, align 4
  %2 = load i32, ptr %value.addr, align 4
  %sub = sub nsw i32 %2, 1
  %call = call noundef i32 @_Z9factoriali(i32 noundef %sub)
  %mul = mul nsw i32 %1, %call
  br label %cond.end

cond.end:
  %cond = phi i32 [ 1, %cond.true ], [ %mul, %cond.false ]
  ret i32 %cond
}
```
Result:
```ll
; test_opt.ll
define dso_local noundef i32 @_Z9factoriali(i32 noundef %value) {
entry:
  call void @instrument_start()
  %value.addr = alloca i32, align 4
  store i32 %value, ptr %value.addr, align 4
  %0 = load i32, ptr %value.addr, align 4
  %cmp = icmp sle i32 %0, 1
  br i1 %cmp, label %cond.true, label %cond.false

cond.true:
  br label %cond.end

cond.false:
  %1 = load i32, ptr %value.addr, align 4
  %2 = load i32, ptr %value.addr, align 4
  %sub = sub nsw i32 %2, 1
  %call = call noundef i32 @_Z9factoriali(i32 noundef %sub)
  %mul = mul nsw i32 %1, %call
  br label %cond.end

cond.end:
  %cond = phi i32 [ 1, %cond.true ], [ %mul, %cond.false ]
  call void @instrument_end()
  ret i32 %cond
}
```

### CODE REVIEW

**Comment by Kuznetsov-Artyom** [comment]
Windows crashes for some reason, I'll figure it out tomorrow.

File: llvm/labs/lab1/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FILE(GLOB children RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/*)
```
Comment (m-ly4) [inline llvm/labs/lab1/CMakeLists.txt]: Rename dir please to `lab2`.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab1/CMakeLists.txt (hunk 1 above)
Comment (Kuznetsov-Artyom) [reply llvm/labs/lab1/CMakeLists.txt]: @m-ly4 I have corrected what you asked for, please correct me if I misunderstood you

Comment (m-ly4) [reply llvm/labs/lab1/CMakeLists.txt]: You did exactly what I asked for. Thank you.

File: llvm/test/lab2/kuznetsov_artyom/test.ll [hunk 2]
```diff
@@ -0,0 +1,75 @@
+; RUN: opt -passes=instr_func -S %s | FileCheck %s
```
Comment (m-ly4) [inline llvm/test/lab2/kuznetsov_artyom/test.ll]: Could you please load this pass from dynamic library using `opt` option `load-pass-plugin`?

File: llvm/test/lab2/kuznetsov_artyom/test.ll [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+; CHECK-LABEL: @_Z8voidFuncv
+; CHECK: call void @instrument_start()
+; CHECK: call void @instrument_end()
+
+define dso_local void @_Z8voidFuncv() {
+entry:
+  ret void
+}
+
```
Comment (m-ly4) [inline llvm/test/lab2/kuznetsov_artyom/test.ll:79]: Please, add a check that there is no new insertions into already instrumented function.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab2/kuznetsov_artyom/test.ll (hunk 3 above)
Comment (Kuznetsov-Artyom) [reply llvm/test/lab2/kuznetsov_artyom/test.ll:79]: > Please, add a check that there is no new insertions into already instrumented function.

Oh, thanks. Indeed, my plugin did not work correctly and the insertion was duplicated. I'll fix it.

Comment (Kuznetsov-Artyom) [reply llvm/test/lab2/kuznetsov_artyom/test.ll:79]: @m-ly4 I've fixed everything.

**Review by m-ly4 — APPROVED** [review]
Thank you for patient applying remarks)
Very good.

**Pull request merged**